File: rust/crates/dw-config/src/doctor.rs
use crate::json::{read_json_value, read_jsonc_value};
use crate::settings::{resolve_root, user_settings_path};
use crate::types::{ConfigDoctorCheck, ConfigDoctorReport, ConfigShow};
use std::path::Path;

pub fn config_show(explicit_root: Option<&str>) -> ConfigShow {
    let root = resolve_root(explicit_root);
    let settings = crate::settings::load_user_settings();
    let workflow_path = Path::new(&root).join("config").join("workflow.json");
    let projects_path = Path::new(&root).join("config").join("projects.json");
    let databases_path = Path::new(&root).join("config").join("databases.json");

    ConfigShow {
        root,
        color: crate::settings::normalize_color_mode(settings.color.as_deref())
            .unwrap_or_else(|_| "auto".into()),
        settings_path: user_settings_path(),
        workflow_path: workflow_path.display().to_string(),
        projects_path: projects_path.display().to_string(),
        databases_path: databases_path.display().to_string(),
        workflow_exists: workflow_path.exists(),
        projects_exists: projects_path.exists(),
        databases_exists: databases_path.exists(),
    }
}

pub fn config_doctor(explicit_root: Option<&str>) -> ConfigDoctorReport {
    let root = resolve_root(explicit_root);
    let checks = vec![
        check_known_config(
            &Path::new(&root).join("config").join("projects.json"),
            &["schema", "projects"],
        ),
        check_known_config(
            &Path::new(&root).join("config").join("workflow.json"),
            &["schema", "branchPrefixes", "azureDevOps", "auth", "updates"],
        ),
        check_known_config(
            &Path::new(&root).join("config").join("databases.json"),
            &["schema", "defaults", "globals", "projects"],
        ),
        check_jsonc(&Path::new(&root).join("config/opencode/opencode.jsonc")),
        check_exists(&Path::new(&root).join("schemas/projects.schema.json")),
        check_exists(&Path::new(&root).join("schemas/workflow.schema.json")),
        check_exists(&Path::new(&root).join("schemas/databases.schema.json")),
    ];
    let passed = checks.iter().all(|check| check.passed);
    ConfigDoctorReport {
        root,
        checks,
        passed,
    }
}

fn check_known_config(path: &Path, required_properties: &[&str]) -> ConfigDoctorCheck {
    let json = match read_json_value(path) {
        Ok(value) => value,
        Err(message) => return doctor_check(path, false, Some(message)),
    };
    let Some(object) = json.as_object() else {
        return doctor_check(
            path,
            false,
            Some("la racine doit etre un objet JSON".into()),
        );
    };
    let missing = required_properties
        .iter()
        .filter(|property| !object.contains_key(**property))
        .copied()
        .collect::<Vec<_>>();
    if missing.is_empty() {
        doctor_check(path, true, None)
    } else {
        doctor_check(
            path,
            false,
            Some(format!("proprietes manquantes: {}", missing.join(", "))),
        )
    }
}

fn check_jsonc(path: &Path) -> ConfigDoctorCheck {
    match read_jsonc_value(path) {
        Ok(_) => doctor_check(path, true, None),
        Err(message) => doctor_check(path, false, Some(message)),
    }
}

fn check_exists(path: &Path) -> ConfigDoctorCheck {
    doctor_check(
        path,
        path.exists(),
        if path.exists() {
            None
        } else {
            Some("fichier manquant".into())
        },
    )
}

fn doctor_check(path: &Path, passed: bool, message: Option<String>) -> ConfigDoctorCheck {
    ConfigDoctorCheck {
        path: path.display().to_string(),
        passed,
        message,
    }
}

File: internal/config/doctor.go
package config

import (
	"os"
	"path/filepath"
	"strings"

	"github.com/sachahjkl/dw/internal/l10n"
	"github.com/sachahjkl/dw/internal/wirejson"
)

func Show(explicitRoot string) ConfigShow {
	root := ResolveRoot(explicitRoot)
	settings := LoadUserSettings()
	workflowPath := filepath.Join(root, "config", "workflow.json")
	projectsPath := filepath.Join(root, "config", "projects.json")
	databasesPath := filepath.Join(root, "config", "databases.json")
	return ConfigShow{
		Root: root, Color: NormalizeColorMode(settings.Color), SettingsPath: UserSettingsPath(),
		WorkflowPath: workflowPath, ProjectsPath: projectsPath, DatabasesPath: databasesPath,
		WorkflowExists: pathExists(workflowPath), ProjectsExists: pathExists(projectsPath),
		DatabasesExists: pathExists(databasesPath),
	}
}

func Status(explicitRoot string) RootStatus {
	root := ResolveRoot(explicitRoot)
	required := []string{
		filepath.Join(root, "config", "projects.json"),
		filepath.Join(root, "config", "workflow.json"),
		filepath.Join(root, "config", "databases.json"),
	}
	missing := make([]string, 0)
	for _, path := range required {
		info, err := os.Stat(path)
		if err != nil || !info.Mode().IsRegular() {
			missing = append(missing, path)
		}
	}
	return RootStatus{Root: root, Initialized: len(missing) == 0, MissingPaths: missing}
}

func Doctor(explicitRoot string) ConfigDoctorReport {
	root := ResolveRoot(explicitRoot)
	checks := []ConfigDoctorCheck{
		checkKnownConfig(filepath.Join(root, "config", "projects.json"), []string{"schema", "projects"}),
		checkKnownConfig(filepath.Join(root, "config", "workflow.json"), []string{"schema", "branchPrefixes", "azureDevOps", "auth", "updates"}),
		checkKnownConfig(filepath.Join(root, "config", "databases.json"), []string{"schema", "defaults", "globals", "projects"}),
		checkJSONC(filepath.Join(root, "config", "opencode", "opencode.jsonc")),
		checkExists(filepath.Join(root, "schemas", "projects.schema.json")),
		checkExists(filepath.Join(root, "schemas", "workflow.schema.json")),
		checkExists(filepath.Join(root, "schemas", "databases.schema.json")),
	}
	passed := true
	for _, check := range checks {
		if !check.Passed {
			passed = false
			break
		}
	}
	return ConfigDoctorReport{Root: root, Checks: checks, Passed: passed}
}

func checkKnownConfig(path string, required []string) ConfigDoctorCheck {
	data, err := os.ReadFile(path)
	if err != nil {
		return doctorCheck(path, false, messageText("config.missing_file"))
	}
	value, err := wirejson.Parse(data)
	if err != nil {
		message := err.Error()
		return doctorCheck(path, false, &message)
	}
	if value.Kind() != wirejson.Object {
		return doctorCheck(path, false, messageText("config.root_json_object"))
	}
	missing := make([]string, 0)
	for _, property := range required {
		if _, ok := value.Lookup(property); !ok {
			missing = append(missing, property)
		}
	}
	if len(missing) == 0 {
		return doctorCheck(path, true, nil)
	}
	message := l10n.Render(l10n.M("config.missing_properties", l10n.A("properties", strings.Join(missing, ", "))))
	return doctorCheck(path, false, &message)
}

func checkJSONC(path string) ConfigDoctorCheck {
	data, err := os.ReadFile(path)
	if err != nil {
		return doctorCheck(path, false, messageText("config.missing_file"))
	}
	_, err = wirejson.Parse(stripJSONCComments(data))
	if err == nil {
		return doctorCheck(path, true, nil)
	}
	message := err.Error()
	return doctorCheck(path, false, &message)
}

func checkExists(path string) ConfigDoctorCheck {
	if pathExists(path) {
		return doctorCheck(path, true, nil)
	}
	return doctorCheck(path, false, messageText("config.missing_file"))
}

func doctorCheck(path string, passed bool, message *string) ConfigDoctorCheck {
	return ConfigDoctorCheck{Path: path, Passed: passed, Message: message}
}

func messageText(id l10n.ID) *string { value := l10n.Text(id); return &value }

func pathExists(path string) bool { _, err := os.Stat(path); return err == nil }

func stripJSONCComments(input []byte) []byte {
	output := make([]byte, 0, len(input))
	inString, escaped := false, false
	for index := 0; index < len(input); index++ {
		char := input[index]
		if inString {
			output = append(output, char)
			if escaped {
				escaped = false
			} else if char == '\\' {
				escaped = true
			} else if char == '"' {
				inString = false
			}
			continue
		}
		if char == '"' {
			inString = true
			output = append(output, char)
			continue
		}
		if char == '/' && index+1 < len(input) && input[index+1] == '/' {
			index += 2
			for index < len(input) && input[index] != '\n' {
				index++
			}
			if index < len(input) {
				output = append(output, '\n')
			}
			continue
		}
		output = append(output, char)
	}
	return output
}
